047-TrafficControlWithDapr/Student/Challenge-03.md

```diff
@@ -0,0 +1,64 @@
+# Challenge 3 - Dapr Pub/Sub Messaging
+
+[< Previous Challenge](./Challenge-02.md) - **[Home](../README.md)** - [Next Challenge >](./Challenge-04.md)
+
+## Introduction
+
+In this challenge, you're going to add Dapr **publish/subscribe** messaging to send messages from the `TrafficControlService` to the `FineCollectionService`.
+
+## Description
+
+In challenge 2, you implemented direct, synchronous communication between two microservices. This pattern is common when an immediate response is required. Communication between services doesn't always require an immediate response.
+
+The publish/subscribe pattern allows your microservices to communicate asynchronously with each other purely by sending messages. In this system, the producer of a message sends it to a topic, with no knowledge of what service(s) will consume the message. A message can even be sent if there's no consumer for it.
+
+Similarly, a subscriber or consumer will receive messages from a topic without knowledge of what producer sent it. This pattern is especially useful when you need to decouple microservices from one another. See the diagram below for an overview of how this pattern works with Dapr:
+
+<img src="../images/Challenge-03/pub-sub.png" style="zoom: 66%;padding-top: 50px;" />
+
+You will need to modify the services to use the Dapr pub/sub building block.
+
+- Start up a RabbitMQ pub/sub message broker in a Docker container.
+- Modify the `TrafficControlService` (`TrafficController` class) so it sends `SpeedingViolation` messages using the Dapr pub/sub building block.
+- Modify the `FineCollectionService` (`CollectionController` class) so it receives `SpeedingViolation` messages using the Dapr pub/sub building block.
+- Create a Dapr configuration file for specifying the pub/sub Dapr components.
+- Restart all services & run the **Simulation** application.
+- Once you have the above working, replace the RabbitMQ message broker with Azure Service Bus without code changes.
+
+## Success Criteria
+
+This challenge targets the operations labeled as **number 2** in the end-state setup:
+
+<img src="../images/Challenge-03/dapr-setup-assignment03.png" style="zoom: 67%;" />
+
+- Validate that the RabbitMQ message broker is running.
+- Validate that messages are being sent from the `TrafficControlService`. to the `FineCollectionService` using Dapr, not direct service invocation.
+- Validate that messages are being received & consumed from the Azure Service Bus.
+
+## Tips
```

Comment:
Are you giving away too much in the tips here?  Should any of this be held back in the coach guide?

I know we discussed that these details are not core to the lesson, so that's why we'd give them to the students. I'm okay with this if you are.  One option to consider when designing a hack is to put details in the coach guide with a note indicating the coach should use their discretion in sharing the detailed steps with students.

Reply:
I felt that given you have to specify certain ports & certain images to pull from Docker Hub (for this one and the SMTP one) that it was distracting from the point of the challenge & would slow things down too much